File: src/Skia/Avalonia.NativeGraphics/src/Avalonia.NativeGraphics.Backend/AvgDrawingContext.cs
using System;
using Avalonia.Media;
using Avalonia.Media.Imaging;
using Avalonia.Native.Interop;
using Avalonia.Platform;
using Avalonia.Rendering.SceneGraph;
using Avalonia.Utilities;

namespace Avalonia.NativeGraphics.Backend
{
    class AvgDrawingContext : IDrawingContextImpl, IDrawingContextWithAcrylicLikeSupport
    {
        private readonly IAvgDrawingContext _native;
        private Matrix _transform;
        private readonly Matrix? _postTransform;
        
        private PlatformRenderInterface _renderInterface;

        public AvgDrawingContext(IAvgDrawingContext native)
        {
            var renderInterace = AvaloniaLocator.Current.GetService<IPlatformRenderInterface>();
            _renderInterface = (PlatformRenderInterface) renderInterace;
            _native = native;

            _postTransform =
                Matrix.CreateScale(_native.Scaling, _native.Scaling);

        }

        public void Dispose() => _native.Dispose();

        public void Clear(Color color)
        {
            _native.Clear(color.ToUint32());
        }
        
        public void DrawBitmap(IRef<IBitmapImpl> source, double opacity, Rect sourceRect, Rect destRect,
            BitmapInterpolationMode bitmapInterpolationMode = BitmapInterpolationMode.Default)
        {
            ImmutableBitmap bmp = (ImmutableBitmap) source.Item;
            
            _native.DrawImage(bmp._image);
        }

        public void DrawBitmap(IRef<IBitmapImpl> source, IBrush opacityMask, Rect opacityMaskRect, Rect destRect)
        {
            
        }

        public void DrawImage(IAvgImage source)
        {
           _native.DrawImage(source); 
        }

        public void DrawLine(IPen pen, Point p1, Point p2)
        {
            _native.DrawLine(new AvgPoint { X = p1.X, Y = p1.Y}, new AvgPoint{X = p2.X, Y = p2.Y}, CreateAvgPen(pen));
        }

        public void DrawGeometry(IBrush brush, IPen pen, IGeometryImpl geometry)
        {
            var impl = (GeometryImpl) geometry;
            _native.DrawGeometry(impl._avgPath, CreateAvgBrush(brush), CreateAvgPen(pen));
        }

        public void DrawRectangle(IExperimentalAcrylicMaterial material, RoundedRect rect)
        {
            if (rect.Rect.Height <= 0 || rect.Rect.Width <= 0)
                return;
            
            AvgRoundRect avgRoundRect = rect.ToAvgRoundRect();

            if (material != null)
            {
                Brush b = new SolidColorBrush(material.TintColor);
                unsafe
                {
                    _native.DrawRectangle(avgRoundRect, CreateAvgBrush(b), CreateAvgPen(null), null, 0);
                }
            }
        }

        private AvgBrush CreateAvgBrush(IBrush brush)
        {
            if (brush == null)
            {
                return new AvgBrush {Valid = 0};
            }

           return brush.ToAvgBrush();
        }

        private AvgPen CreateAvgPen(IPen pen)
        {
            if (pen == null)
            {
                return new AvgPen {Valid = 0};
            }

            return pen.ToAvgPen();
        }
        
        public void DrawRectangle(IBrush brush, IPen pen, RoundedRect rect, BoxShadows boxShadows = new BoxShadows())
        {
            AvgRoundRect avgRoundRect = rect.ToAvgRoundRect();

            AvgBoxShadow[] bs = new AvgBoxShadow[boxShadows.Count];
            int idx = 0;
            foreach (var boxShadow in boxShadows)
            {
                bs[idx++] = new AvgBoxShadow
                {
                    OffsetX = boxShadow.OffsetX,
                    OffsetY = boxShadow.OffsetY,
                    Blur = boxShadow.Blur,
                    Spread = boxShadow.Spread,
                    color = (int) boxShadow.Color.ToUint32(),
                    IsInset = Convert.ToInt32(boxShadow.IsInset),
                };
            }
            
            unsafe
            {
                fixed (AvgBoxShadow* b = bs)
                {
                    _native.DrawRectangle(avgRoundRect, CreateAvgBrush(brush), CreateAvgPen(pen), b, boxShadows.Count);
                }
            }
        }
        
        public void DrawEllipse(IBrush brush, IPen pen, Rect rect)
        {
            throw new NotImplementedException();
        }

        public void DrawGlyphRun(IBrush foreground, GlyphRun glyphRun)
        {
            //Console.WriteLine($"Draw Glyphrun! - {glyphRun.GlyphRunImpl}");
            GlyphRunImpl glyphRunImpl = (GlyphRunImpl) glyphRun.GlyphRunImpl;
            _native.DrawGlyphRun(glyphRunImpl.AvgGlyphRun, glyphRun.BaselineOrigin.X, glyphRun.BaselineOrigin.Y, foreground.ToAvgBrush());
        }

        public IDrawingContextLayerImpl CreateLayer(Size size)
        {
            throw new NotImplementedException();
        }

        public void PushClip(Rect clip)
        {
           PushClip(new RoundedRect(clip));
        }

        public void PushClip(RoundedRect clip)
        {
            var avgRect = clip.ToAvgRoundRect();
            _native.PushClip(avgRect); 
        }

        public void PopClip()
        {
            _native.PopClip();
        }

        public void PushOpacity(double opacity)
        {
            _native.PushOpacity(opacity);
        }

        public void PopOpacity()
        {
            _native.PopOpacity();
        }

        public void PushOpacityMask(IBrush mask, Rect bounds)
        {
            
        }

        public void PopOpacityMask()
        {
            
        }

        public void PushGeometryClip(IGeometryImpl clip)
        {
            
        }

        public void PopGeometryClip()
        {
            
        }

        public void PushBitmapBlendMode(BitmapBlendingMode blendingMode)
        {
            
        }

        public void PopBitmapBlendMode()
        {
            
        }

        public void Custom(ICustomDrawOperation custom)
        {
            
        }

        public unsafe Matrix Transform
        {
            get => _transform;
            set
            {
                var transform = value;
                
                if (_postTransform.HasValue)
                {
                    transform *= _postTransform.Value;
                }

                _transform = value;
                
                var m = new AvgMatrix3x2
                {
                    M11 = transform.M11,
                    M12 = transform.M12,
                    M21 = transform.M21,
                    M22 = transform.M22,
                    M31 = transform.M31,
                    M32 = transform.M32,
                };
                
                _native.SetTransform(&m);
            }
        }
    }
}

File: src/Skia/Avalonia.NativeGraphics/native/src/AvgDrawingContext.cpp
//
// Created by kekekeks on 04.05.22.
//

#include "AvgDrawingContext.h"
#include "AvgPath.h"
#include "AvgImage.h"
#include "AvgGlyphRun.h"
#include "comimpl.h"
#include "include/core/SkRRect.h"
#include "interop.h"

void AvgDrawingContext::SetTransform(AvgMatrix3x2* matrix)
{
    SkMatrix skMatrix;
    skMatrix.setAll((float)matrix->M11, (float)matrix->M21, (float)matrix->M31, (float)matrix->M12, (float)matrix->M22,
                    (float)matrix->M32, 0, 0, 1);
    _canvas->setMatrix(skMatrix);
}

void AvgDrawingContext::Clear(unsigned int color)
{
    _canvas->clear(color);
}

void AvgDrawingContext::DrawImage(IAvgImage* image)
{
    AvgImage* avgImage = dynamic_cast<AvgImage*>(image);

    _canvas->drawImage(avgImage->_image, 0, 0);
}

void AvgDrawingContext::ConfigurePaint(AvgBrush brush)
{
    double opacity = brush.Opacity * _currentOpacity;

    SkColor color = SkColorSetARGB(((double)(brush.Color.A) * opacity), brush.Color.R, brush.Color.G, brush.Color.B);
    _fillPaint.reset();
    _fillPaint.setColor(color);
    _fillPaint.setAntiAlias(true);
}

void AvgDrawingContext::ConfigurePaint(AvgPen pen)
{
    _strokePaint.reset();
    if (pen.Thickness == 0.0)
    {
        return;
    }

    AvgBrush& brush = pen.Brush;

    double opacity = brush.Opacity * _currentOpacity;
    SkColor color = SkColorSetARGB(((double)(brush.Color.A) * opacity), brush.Color.R, brush.Color.G, brush.Color.B);
    _strokePaint.setColor(color);
    _strokePaint.setAntiAlias(true);
    _strokePaint.setStroke(true);
    _strokePaint.setStrokeWidth(pen.Thickness);
}

void AvgDrawingContext::DrawGeometry(IAvgPath* path, AvgBrush brush, AvgPen pen)
{
    AvgPath* avgPath = dynamic_cast<AvgPath*>(path);

    if (brush.Valid)
    {
        ConfigurePaint(brush);
        _canvas->drawPath(avgPath->_path, _fillPaint);
    }

    if (pen.Valid)
    {
        ConfigurePaint(pen);
        _canvas->drawPath(avgPath->_path, _strokePaint);
    }
}

void AvgDrawingContext::DrawRectangle(AvgRoundRect rrect, AvgBrush brush, AvgPen pen, AvgBoxShadow* boxshadows,
                                      int n_boxshadows)
{
    AvgRect rect = rrect.Rect;
    SkRect skRect = SkRect::MakeXYWH(rect.X, rect.Y, rect.X + rect.Width, rect.Y + rect.Height);
    SkRRect skRRect;

    SkVector corners[] = {
        {(float)rrect.RadiiTopLeft.X, (float)rrect.RadiiTopLeft.Y},
        {(float)rrect.RadiiTopRight.X, (float)rrect.RadiiTopRight.Y},
        {(float)rrect.RadiiBottomRight.X, (float)rrect.RadiiBottomRight.Y},
        {(float)rrect.RadiiBottomLeft.X, (float)rrect.RadiiBottomLeft.Y},
    };

    if (rrect.IsRounded)
    {
        skRRect.setRectRadii(skRect, corners);
    }

    for (int i = 0; i < n_boxshadows; i++)
    {
        printf("Point %d: {%f,%f}\n", i, boxshadows[i].OffsetX, boxshadows[i].OffsetY);
    }

    if (brush.Valid)
    {
        ConfigurePaint(brush);
        if (rrect.IsRounded)
        {
            _canvas->drawRRect(skRRect, _fillPaint);
        }
        else
        {
            _canvas->drawRect(skRect, _fillPaint);
        }
    }

    if (pen.Valid)
    {
        ConfigurePaint(pen);
        if (rrect.IsRounded)
        {
            _canvas->drawRRect(skRRect, _strokePaint);
        }
        else
        {
            _canvas->drawRect(skRect, _strokePaint);
        }
    }
}

void AvgDrawingContext::DrawLine(AvgPoint p1, AvgPoint p2, AvgPen pen)
{
    if (pen.Valid)
    {
        ConfigurePaint(pen);
        _canvas->drawLine(p1.X, p1.Y, p2.X, p2.Y, _strokePaint);
    }
}

void AvgDrawingContext::DrawGlyphRun(IAvgGlyphRun* glyphRun, double x, double y, AvgBrush brush)
{
    AvgGlyphRun* avgGlyphRun = dynamic_cast<AvgGlyphRun*>(glyphRun);

    ConfigurePaint(brush);

    // printf("Calling drawTextBlob() %f,%f : %f:%f\n", x,y, avgGlyphRun->_textBlob.get()->bounds().x(), avgGlyphRun->_textBlob.get()->bounds().y());

    _canvas->drawTextBlob(avgGlyphRun->_textBlob, x, y, _fillPaint);
    // _canvas->drawSimpleText("hello world", 10, SkTextEncoding::kUTF8, x,y, avgGlyphRun->_fontManager->_font, _fillPaint);
}

void AvgDrawingContext::PushOpacity(double opacity)
{
    _opacityStack.push(_currentOpacity);
    _currentOpacity *= opacity;
}

void AvgDrawingContext::PopOpacity()
{
    _currentOpacity = _opacityStack.top();
    _opacityStack.pop();
}

void AvgDrawingContext::PushClip(AvgRoundRect clip)
{
    AvgRect rect = clip.Rect;
    SkRect skRect = {(float)rect.X, (float)rect.Y, (float)(rect.X + rect.Width), (float)(rect.Y + rect.Height)};
    SkRRect skRRect;
    SkVector corners[] = {
        {(float)clip.RadiiTopLeft.X, (float)clip.RadiiTopLeft.Y},
        {(float)clip.RadiiTopRight.X, (float)clip.RadiiTopRight.Y},
        {(float)clip.RadiiBottomRight.X, (float)clip.RadiiBottomRight.Y},
        {(float)clip.RadiiBottomLeft.X, (float)clip.RadiiBottomLeft.Y},
    };

    _canvas->save();

    if (clip.IsRounded)
    {
        skRRect.setRectRadii(skRect, corners);
        _canvas->clipRRect(skRRect);
    }
    else
    {
        _canvas->clipRect(skRect);
    }
}

void AvgDrawingContext::PopClip()
{
    _canvas->restore();
}

double AvgDrawingContext::GetScaling()
{
    return _scaling;
}
